### PR TITLE
Overriding CMake Release

### PR DESCRIPTION
**Pull request by jerinphilip**
### Description

Release is now Slim - NDEBUG. RELWITHDEBINFO was previously equivalent
to Release, which is no longer the case. 

https://github.com/browsermt/bergamot-translator/pull/175#issuecomment-847810390

Added dependencies: none

### How to test
Describe how to test your changes, adding command line examples and sample input/output files if relevant.
Point to unit tests or regression tests covering the changes if they have been added.

Describe how you have tested your code, including OS and the cmake command.

<details>
<summary>Description</summary>

```
jphilip@var:~/code/marian-dev/build$ cmake .. -DCMAKE_BUILD_TYPE=Release -DCOMPILE_CUDA=off
-- Project name: marian
-- Project version: v1.9.56+c0553d76
fatal: reference is not a tree: 3ffdc0065a03cadd9d0e5e123aaf9b6ea7ffb05d
Unable to checkout '3ffdc0065a03cadd9d0e5e123aaf9b6ea7ffb05d' in submodule path 'src/3rd_party/sentencepiece'
-- Checking support for CPU intrinsics
-- SSE2 support found
-- SSE3 support found
-- SSE4.1 support found
-- SSE4.2 support found
-- AVX support found
-- AVX2 support found
-- AVX512 support found
CMake Warning at CMakeLists.txt:463 (message):
  COMPILE_CUDA=off : Building only CPU version


-- Found Tcmalloc: /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so
-- VERSION: 0.1.94
-- Found TCMalloc: /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so
-- Configuring done
-- Generating done
-- Build files have been written to: /home/jphilip/code/marian-dev/build
jphilip@var:~/code/marian-dev/build$ make -j &> error.log || vim error.log
jphilip@var:~/code/marian-dev/build$ du -h marian*
46M     marian
39M     marian-conv
43M     marian-decoder
43M     marian-scorer
5.4M    marian-vocab
```

</details>

### Checklist

- [ ] I have tested the code manually
- [ ] I have run regression tests
- [ ] I have read and followed CONTRIBUTING.md
- [ ] I have updated CHANGELOG.md

### CODE REVIEW

**Comment by XapaJIaMnu** [comment]
I want to test this a bit more on a native windows environment, because I seem to recall some weird interaction with debug flags on windows.

File: CMakeLists.txt [hunk 1]
```diff
@@ -274,23 +275,23 @@ else(MSVC)
...
-    set(CMAKE_CXX_FLAGS_PROFGEN         "${CMAKE_CXX_FLAGS_RELEASE} -fprofile-generate -fprofile-correction")
-    set(CMAKE_CXX_FLAGS_PROFUSE         "${CMAKE_CXX_FLAGS_RELEASE} -fprofile-use -fprofile-correction")
+    set(CMAKE_CXX_FLAGS_RELWITHDEBINFO  "${CMAKE_CXX_FLAGS_RELEASE} -g ${CMAKE_RDYNAMIC_FLAG}")
+    set(CMAKE_CXX_FLAGS_PROFILE         "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -pg")
+    set(CMAKE_CXX_FLAGS_PROFGEN         "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -fprofile-generate -fprofile-correction")
+    set(CMAKE_CXX_FLAGS_PROFUSE         "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -fprofile-use -fprofile-correction")
 
     # these need to be set separately
     set(CMAKE_C_FLAGS                 "-pthread ${CMAKE_GCC_FLAGS} -fPIC ${DISABLE_GLOBALLY} -march=${BUILD_ARCH} ${INTRINSICS}")
-    set(CMAKE_C_FLAGS_RELEASE         "-O3 ${BUILD_WIDTH} -funroll-loops -g ${CMAKE_RDYNAMIC_FLAG}")
```
Comment (XapaJIaMnu) [inline CMakeLists.txt:287]: Why did the RDYNAMIC flag disappear?

File: CMakeLists.txt [hunk 2]
```diff
@@ -274,23 +275,23 @@ else(MSVC)
                                               --shell-file ${CMAKE_SOURCE_DIR}/wasm/custom_shell.html")
   else(COMPILE_WASM)
     set(CMAKE_CXX_FLAGS                 "-std=c++11 -pthread ${CMAKE_GCC_FLAGS} -fPIC ${DISABLE_GLOBALLY} -march=${BUILD_ARCH} ${INTRINSICS} ${BUILD_WIDTH}")
-    set(CMAKE_CXX_FLAGS_RELEASE         "-O3 -funroll-loops -g ${CMAKE_RDYNAMIC_FLAG}")
```
Comment (XapaJIaMnu) [inline CMakeLists.txt:277]: RDYNAMIC flag?

File: CMakeLists.txt (hunk 1 above)
Comment (jerinphilip) [reply CMakeLists.txt:287]: I'm no expert, but googling gave me the idea this is a debug related flag?

> -rdynamic
> Pass the flag -export-dynamic to the ELF linker, on targets that support it. This instructs the linker to 
 add all symbols, not only used ones, to the dynamic symbol table. This option is needed for some uses of dlopen or to allow obtaining backtraces from within a program.
 
 Do we need this in Release?

Comment (XapaJIaMnu) [reply CMakeLists.txt:287]: Not only, https://stackoverflow.com/questions/36692315/what-exactly-does-rdynamic-do-and-when-exactly-is-it-needed . I guess if compilation succeeds without one, it's probably fine.

File: CMakeLists.txt [hunk 3]
```diff
@@ -274,23 +275,23 @@ else(MSVC)
...
+    set(CMAKE_CXX_FLAGS_RELEASE         "-O3 ${BUILD_WIDTH} -funroll-loops")
     set(CMAKE_CXX_FLAGS_DEBUG           "-O0 -g ${CMAKE_RDYNAMIC_FLAG}")
     set(CMAKE_CXX_FLAGS_SLIM            "-O3 ${BUILD_WIDTH} -funroll-loops -DNDEBUG")
-    set(CMAKE_CXX_FLAGS_RELWITHDEBINFO  "${CMAKE_CXX_FLAGS_RELEASE}")
-    set(CMAKE_CXX_FLAGS_PROFILE         "${CMAKE_CXX_FLAGS_RELEASE} -pg")
-    set(CMAKE_CXX_FLAGS_PROFGEN         "${CMAKE_CXX_FLAGS_RELEASE} -fprofile-generate -fprofile-correction")
-    set(CMAKE_CXX_FLAGS_PROFUSE         "${CMAKE_CXX_FLAGS_RELEASE} -fprofile-use -fprofile-correction")
+    set(CMAKE_CXX_FLAGS_RELWITHDEBINFO  "${CMAKE_CXX_FLAGS_RELEASE} -g ${CMAKE_RDYNAMIC_FLAG}")
```
Comment (XapaJIaMnu) [inline CMakeLists.txt:281]: The `${BUILD_WDITH}` is not consistently enforced, but that's probably my fault, will fix later.

**Review by XapaJIaMnu — APPROVED** [review]

**Pull request merged**